### DISCUSSION TITLE
Additional features for fields

### DISCUSSION THREAD

**Discussion by VaZark** [directus/directus#11424] [Feature Requests]
I've just discovered `directus` by accident after coming across the name on a job posting. So far I'm thoroughly fascinated with the framework/application. However, it seems like there are a few options missing in regards to the fields. Can you please clarify if they do not exist or if I am missing something?

## Locking / Required fields and collections

As a developer, when creating a website for a client that requires a custom endpoint that depends upon a particular field in a collection, I want to guarantee that the fields exist. 

The only way that is possible is if the fields are "locked" and cannot be deleted from the UI without dev intervention. This can essentially be considered a safety feature for customized endpoints.

Something like the below snippet in the base config is what I had in mind.
```js
directus_db_config = {
 required_fields : { "COLLECTION_NAME" : [ "field_1", "field_9" ] }
}
```
This will of course require that the Collection and the field are autogenerated at application startup though. So this feature can be extended to generate Collections based on an initial migration (or templates?) and locking them in place. 

## Soft-delete

Sometimes even if the data is no longer needed for an application to function, it needs to be available for X years due to legal constraints. So an option to hide the column and delete it after X amount of time has passed would be fantastic

**Comment by rijkvanzanten** [comment]
> The only way that is possible is if the fields are "locked" and cannot be deleted from the UI without dev intervention. This can essentially be considered a safety feature for customized endppoints.

Only the admin role has the ability to delete fields. I'd strongly recommend setting up a non-admin role for your non-dev users 👍🏻

**Reply by VaZark** [reply]
True. However, with the lack of soft delete, if the client or even a new dev deletes a field since they no longer need useful, it might crash any customized functions with no way safe way to recover on their own. That is the reason I grouped the two options together in this discussion

**Reply by benhaynes** [reply]
Perhaps you could create a custom hook that locks that schema change based on a deny-list of tables. Could be a whole workflow for configuring this.

**Reply by benhaynes** [reply]
That customizable workflow could also be used to delete fields after X time.

**Reply by VaZark** [reply]
Hmmm… that’s one way to go about that.

How feasible to do you it’ll be to extend the permissions model objects for fields as well ? That’ll ensure delete can be disabled and even the api can fine-tuned accesses based on user roles.

Eg : some users can just only view some fields but can modify the rest. Useful when some states and statuses must be updated based on certain conditions

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
